pkg/controller/clusterpool/clusterpool_controller_test.go

```diff
@@ -0,0 +1,215 @@
+
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/openshift/hive/pkg/apis"
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
+	"github.com/openshift/hive/pkg/apis/hive/v1/aws"
+	hivev1aws "github.com/openshift/hive/pkg/apis/hive/v1/aws"
+)
+
+const (
+	testNamespace     = "test-namespace"
+	hiveNamespace     = "hive"
+	testLeasePoolName = "aws-us-east-1"
+	credsSecretName   = "aws-creds"
+)
+
+func init() {
+	log.SetLevel(log.DebugLevel)
+}
+
+func TestReconcileClusterPool(t *testing.T) {
+	apis.AddToScheme(scheme.Scheme)
+
+	d, err := time.ParseDuration("8h")
+	require.NoError(t, err)
+	eightHourDur := &metav1.Duration{Duration: d}
+
+	tests := []struct {
+		name      string
+		existing  []runtime.Object
+		expectErr bool
+		//validate                func(client.Client, *testing.T)
+		expectedClusters        int
+		expectedDeletedClusters int
+	}{
+		{
+			name: "create all clusters",
+			existing: []runtime.Object{
+				buildPool(5, eightHourDur),
+				buildSecret(corev1.SecretTypeOpaque, hiveNamespace, credsSecretName, "dummykey", "dummyval"),
+			},
+			expectedClusters: 5,
+		},
+		{
+			name: "scale up",
+			existing: []runtime.Object{
+				buildPool(5, eightHourDur),
+				buildSecret(corev1.SecretTypeOpaque, hiveNamespace, credsSecretName, "dummykey", "dummyval"),
+				testClusterDeployment("c1"),
+				testClusterDeployment("c2"),
+				testClusterDeployment("c3"),
+			},
+			expectedClusters: 5,
+		},
+		{
+			name: "scale down",
+			existing: []runtime.Object{
+				buildPool(3, eightHourDur),
+				buildSecret(corev1.SecretTypeOpaque, hiveNamespace, credsSecretName, "dummykey", "dummyval"),
+				testClusterDeployment("c1"),
+				testClusterDeployment("c2"),
+				testClusterDeployment("c3"),
+				testClusterDeployment("c4"),
+				testClusterDeployment("c5"),
+				testClusterDeployment("c6"),
+			},
+			expectedClusters: 3,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			//logger := log.WithField("controller", "clusterProvision")
+			fakeClient := fake.NewFakeClient(test.existing...)
+			rcp := &ReconcileClusterPool{
+				Client: fakeClient,
+				scheme: scheme.Scheme,
+			}
+
+			reconcileRequest := reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      testLeasePoolName,
+					Namespace: testNamespace,
+				},
+			}
+
+			_, err := rcp.Reconcile(reconcileRequest)
+
+			if test.expectErr {
+				assert.Error(t, err, "expected error from reconcile")
+			} else {
+				assert.NoError(t, err, "expected no error from reconcile")
+			}
+
+			cds := &hivev1.ClusterDeploymentList{}
+			err = fakeClient.List(context.Background(), cds)
+			require.NoError(t, err)
+
+			var deletingCDs int
+			for _, cd := range cds.Items {
+				if cd.DeletionTimestamp != nil {
+					deletingCDs++
+				}
+			}
+
+			assert.Equal(t, test.expectedClusters, len(cds.Items))
+			assert.Equal(t, test.expectedDeletedClusters, deletingCDs)
```

Comment:
This is not testing anything. No tests set `expectedDeletedClusters`, and `deletingCDs` will always be 0.